### PR TITLE
Update DataLoader.lua

### PR DESCRIPTION
**Pull request by haooooooqi**

### CODE REVIEW

**Comment by ywelement** [comment]
@takatosp1 , thanks for updating DataLoader functions. Could you update test.lua as well to include a unit test for the new implementation if possible? We'd like to keep this good practice for better quality maintenance.

**Comment by takatosp1** [comment]
Sure wait me work on it

File: DataLoader.lua [hunk 1]
```diff
@@ -5,10 +5,20 @@ function DataLoader:index(indices, inputs, targets, ...)
...
 end
 
+
+
 function DataLoader:sample(batchsize, inputs, targets, ...)
-   self._indices = self._indices or torch.LongTensor()
-   self._indices:resize(batchsize):random(1,self:size())
-   return self:index(self._indices, inputs, targets, ...)
+    self._indices = self._indices or torch.LongTensor()
+    if ifTrain == true then
```
Comment (nicholas-leonard) [inline DataLoader.lua:12]: The ifTrain logic doesn't fit in. For training use sample. For testing use sub or index.

File: DataLoader.lua [hunk 2]
```diff
@@ -18,11 +28,17 @@ function DataLoader:sub(start, stop, inputs, targets, ...)
 end
 
 function DataLoader:shuffle()
-   error"Not Implemented"
+   self._permuteIndex = torch.randperm(self.imagePath:size(1)):long()
 end
 
 function DataLoader:split(ratio)
```
Comment (nicholas-leonard) [inline DataLoader.lua:34]: This implementation won't work for all sub-classes. `imagePath` is only supported by `ImageClass`.

**Review by nicholas-leonard — COMMENTED** [review]
I recommend that this PR shouldn't be merged. It replicates functionality that already exists, or should be added in the appropriate sub-classes.

**Comment by ywpre** [comment]
@nicholas-leonard , thanks for the recommendation! You're right, it makes sense to keep the base class general rather than covering sub-cases.

**Comment by ywpre** [comment]
@takatosp1, would it be OK to leave your changes separate from the main branch such that it won't cause any potential conflicts with sub-classes? Please let us know if you have any other suggestions.

**Pull request closed without merging**